File: src/python/bfs.py
import json
import sys
from collections import deque

class Graph:
    def __init__(self):
        self.adj_list = {}

    def add_edge(self, src, dest):
        if src not in self.adj_list:
            self.adj_list[src] = []
        self.adj_list[src].append(dest)

    def bfs(self, start_node):
        queue = deque([start_node])
        visited = set([start_node])
        bfs_order = []

        while queue:
            node = queue.popleft()
            bfs_order.append(node)

            for neighbor in self.adj_list.get(node, []):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)

        return bfs_order

    @staticmethod
    def read_graph_from_file(file_path):
        with open(file_path, 'r') as file:
            data = json.load(file)
            graph_data = data["graph"]
            start_node = data["start_node"]

            graph = Graph()
            for src, neighbors in graph_data.items():
                for dest in neighbors:
                    graph.add_edge(src, dest)

            return graph, start_node

    @staticmethod
    def write_bfs_order_to_file(file_path, bfs_order):
        output_data = {"bfs_order": bfs_order}
        with open(file_path, 'w') as file:
            json.dump(output_data, file, indent=4)

def main():
    if len(sys.argv) != 3:
        print("Usage: python bfs.py <input_file> <output_file>")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = sys.argv[2]

    # Read graph and start node from input file
    graph, start_node = Graph.read_graph_from_file(input_file)

    # Perform BFS
    bfs_order = graph.bfs(start_node)

    # Write the BFS traversal order to the output file
    Graph.write_bfs_order_to_file(output_file, bfs_order)

    print("BFS traversal completed.")

if __name__ == "__main__":
    main()

File: src/typescript/bfs.ts
import * as fs from 'fs';

class Queue<T> {
    private items: T[] = [];

    enqueue(item: T): void {
        this.items.push(item);
    }

    dequeue(): T | undefined {
        return this.items.shift();
    }

    isEmpty(): boolean {
        return this.items.length === 0;
    }
}

class Graph {
    private adjList: { [key: string]: string[] } = {};

    addEdge(src: string, dest: string): void {
        if (!this.adjList[src]) {
            this.adjList[src] = [];
        }
        this.adjList[src].push(dest);
    }

    bfs(startNode: string): string[] {
        const queue = new Queue<string>();
        const visited = new Set<string>();
        const bfsOrder: string[] = [];

        queue.enqueue(startNode);
        visited.add(startNode);

        while (!queue.isEmpty()) {
            const node = queue.dequeue()!;
            bfsOrder.push(node);

            if (this.adjList[node]) {
                for (const neighbor of this.adjList[node]) {
                    if (!visited.has(neighbor)) {
                        visited.add(neighbor);
                        queue.enqueue(neighbor);
                    }
                }
            }
        }

        return bfsOrder;
    }

    static readGraphFromFile(filePath: string): { graphInstance: Graph, startNode: string } {
        const data = fs.readFileSync(filePath, 'utf8');
        const { graph, start_node: startNode } = JSON.parse(data);
        const graphInstance = new Graph();

        for (const src in graph) {
            for (const dest of graph[src]) {
                graphInstance.addEdge(src, dest);
            }
        }

        return { graphInstance, startNode };
    }

    static writeBFSOrderToFile(filePath: string, bfsOrder: string[]): void {
        const outputData = { bfs_order: bfsOrder };
        fs.writeFileSync(filePath, JSON.stringify(outputData, null, 4), 'utf8');
    }
}

// Main function to perform BFS using the Graph class
function main() {
    if (process.argv.length !== 4) {
        console.error("Usage: ts-node bfs.ts <input_file> <output_file>");
        process.exit(1);
    }

    const inputFilePath = process.argv[2];
    const outputFilePath = process.argv[3];

    // Read graph and start node from input file
    const { graphInstance, startNode } = Graph.readGraphFromFile(inputFilePath);

    // Perform BFS
    const bfsOrder = graphInstance.bfs(startNode);

    // Write the BFS traversal order to the output file
    Graph.writeBFSOrderToFile(outputFilePath, bfsOrder);

    console.log("BFS traversal completed.");
}

// Run the main function
main();
